providers/adfs.go

```diff
@@ -0,0 +1,81 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"net/url"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+)
+
+// ADFSProvider represents an ADFS based Identity Provider
+type ADFSProvider struct {
+	*OIDCProvider
+}
+
+var _ Provider = (*OIDCProvider)(nil)
+
+const (
+	adfsProviderName = "Adfs"
+	adfsDefaultScope = "openid email profile"
+)
+
+// NewADFSProvider initiates a new AdfsProvider
+func NewADFSProvider(p *OIDCProvider) *ADFSProvider {
+
+	defaultScope := adfsDefaultScope
+
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		resource := p.ProtectedResource.String()
+		if !strings.HasSuffix(resource, "/") {
+			resource += "/"
+		}
+
+		if p.Scope != "" && !strings.HasPrefix(p.Scope, resource) {
+			p.Scope = resource + p.Scope
+		}
+		defaultScope = resource + defaultScope
+	}
+
+	p.ProviderData.setProviderDefaults(providerDefaults{
+		name:  adfsProviderName,
+		scope: defaultScope,
+	})
+
+	return &ADFSProvider{OIDCProvider: p}
+}
+
+// GetLoginURL Override to double encode the state parameter
+func (p *ADFSProvider) GetLoginURL(redirectURI, state string) string {
+	extraParams := url.Values{}
+	a := makeLoginURL(p.ProviderData, redirectURI, url.QueryEscape(state), extraParams)
+	return a.String()
+}
+
+// EnrichSession to add email
+func (p *ADFSProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	if s.Email != "" {
+		return nil
+	}
+
+	idToken, err := p.Verifier.Verify(ctx, s.IDToken)
+	if err != nil {
+		return err
+	}
+
+	p.EmailClaim = "upn"
+	c, err := p.getClaims(idToken)
+
+	s.Email = c.Email
+	if err != nil {
+		return fmt.Errorf("couldn't extract claims from id_token (%v)", err)
+	}
```

Comment:
Let's keep the error handler next to the call and not set anything on the session until we confirm no errors in `getClaims`

```suggestion
	c, err := p.getClaims(idToken)
	if err != nil {
		return fmt.Errorf("couldn't extract claims from id_token (%v)", err)
	}
	s.Email = c.Email
```